File: oauth-authenticator-archetype/src/main/resources/archetype-resources/src/main/java/authentication/CallbackRequestModel.java
package ${package}.authentication;

import se.curity.identityserver.sdk.Nullable;
import se.curity.identityserver.sdk.web.Request;

import java.util.function.Function;

class CallbackRequestModel
{
    @Nullable
    private final String _error;

    @Nullable
    private final String _errorDescription;

    private final String _url;
    private final String _code;
    private final String _state;

    CallbackRequestModel(Request request)
    {
        Function<String, ? extends RuntimeException> invalidParameter = (s) -> new RuntimeException(String.format(
                "Expected only one query string parameter named %s, but found multiple.", s));

        _code = request.getQueryParameterValueOrError("code", invalidParameter);
        _state = request.getQueryParameterValueOrError("state", invalidParameter);
        _error = request.getQueryParameterValueOrError("error", invalidParameter);
        _errorDescription = request.getQueryParameterValueOrError("error_description", invalidParameter);
        _url = request.getUrl();
    }

    public String getCode()
    {
        return _code;
    }

    public String getState()
    {
        return _state;
    }

    @Nullable
    public String getErrorDescription()
    {
        return _errorDescription;
    }

    public String getRequestUrl()
    {
        return _url;
    }

    @Nullable
    public String getError()
    {
        return _error;
    }
}

File: kotlin-oauth-authenticator-archetype/src/main/resources/archetype-resources/src/main/kotlin/authentication/CallbackRequestModel.kt
package ${package}.authentication

import se.curity.identityserver.sdk.web.Request

class CallbackRequestModel(request: Request)
{
    val error: String? = request.getQueryParameterValueOrError("error", invalidParameter)
    val errorDescription: String? = request.getQueryParameterValueOrError("error_description", invalidParameter)
    val url: String = request.url
    val code: String = request.getQueryParameterValueOrError("code", invalidParameter)
    val state: String = request.getQueryParameterValueOrError("state", invalidParameter)

    companion object
    {
        private val invalidParameter = { s: String -> RuntimeException(String.format(
                "Expected only one query string parameter named %s, but found multiple.", s)) }
    }
}
